### DISCUSSION TITLE
DeltaTable.to_pyarrow_dataset()  is much slower than DeltaTable.files()

### DISCUSSION THREAD

**Discussion by adriangb** [delta-io/delta-rs#2484] [General]
I'm seeing ~500ms for the former and ~1ms for the later (running locally, absolute numbers shouldn't matter)

@ion-elgreco is it possible the table state is not being properly re-used or similar?

**Comment by ion-elgreco** [comment]
@adriangb both do quiete different things

.files() returns the filepaths of the table in a list, while to_pyarrow_dataset also has to construct dataset fragments of each file while parsing the table stats into pyarrow.compute expressions

**Reply by adriangb** [reply]
That makes sense but should parsing that stuff be that expensive? I feel like that should be pretty cheap, there must be extra IO going on for the difference to be so large.

**Reply by adriangb** [reply]
I implemented a "hacky" by hand version and on the same system it takes ~30ms, which is more in line with what I'd expect. And this is in Python, doing regexes, etc. It could be done much better.

```
files = dt.files()
# files will have a path like project_id={uuid}/timestamp={YYYY-MM-DD HH:mm:ss.ffffff}/{random string}.parquet
partitions: dict[tuple[str, tuple[str, str]], list[str]] = defaultdict(list)
for file in files:
    match = re.match(r'^project_id=([^\/]+)/timestamp=([^\/]+)\/.*$', file)
    assert match is not None, f'Unexpected file format: {file}'
    project_id, timestamp = cast(tuple[str, str], match.groups())
    timestamp_dt = datetime.datetime.strptime(timestamp, '%Y-%m-%d %H:%M:%S.%f')
    timestamp_start = timestamp_dt.strftime('%Y-%m-%d %H:%M:%S.%f')
    timestamp_end = (timestamp_dt + datetime.timedelta(hours=1)).strftime('%Y-%m-%d %H:%M:%S.%f')
    partitions[(project_id, (timestamp_start, timestamp_end))].append(file)

format = pyarrow.dataset.ParquetFileFormat(
    default_fragment_scan_options=pyarrow.dataset.ParquetFragmentScanOptions(
        buffer_size=1024 * 1024 * 10,  # 10 MiB
        pre_buffer=True,
    )
)
filesystem, path_prefix = pyarrow.fs.FileSystem.from_uri(dt.table_uri)  # type: ignore
fragments = [  # type: ignore
    format.make_fragment(  # type: ignore
        f'{path_prefix}/{file}',
        filesystem=filesystem,
        partition_expression=(  # type: ignore
            (pyarrow.compute.field('project_id') == pyarrow.compute.scalar(project_id))  # type: ignore
            &
            (pyarrow.compute.field('timestamp') >= pyarrow.compute.scalar(timestamp_start))  # type: ignore
            &
            (pyarrow.compute.field('timestamp') < pyarrow.compute.scalar(timestamp_end))  # type: ignore
        )
    )
    for (project_id, (timestamp_start, timestamp_end)), files in partitions.items()
    for file in files
]
```

**Reply by wjones127** [reply]
`to_pyarrow_dataset` does a few things different than what you are doing here:

1. It parses the partition values from the transaction log, not the files paths. This is actually required by the delta lake spec, since the partition values aren't guaranteed to be in the path. (Most writers do this anyways to be nice, but officially you should get the values from the transaction log.)
2. In addition to partition values, it also parses all the file-level column statistics.

I'd recommend building the library in release mode with debug symbols, then using [pyspy](https://github.com/benfred/py-spy?tab=readme-ov-file#usage) to generate a flamegraph. That would probably give some helpful information to see where time is spent.

**Reply by adriangb** [reply]
It doesn't sound like any of that should be this slow. All of that data is already available (right?) so no IO is needed so this is just CPU to loop over partition values, etc. That shouldn't take hundreds of ms.

Consider this self contained example:

```python
from random import choices
import random
import string
from time import time
from uuid import uuid4
import deltalake
import polars as pl

n = 1_000

df = pl.DataFrame(
    {
        'a': [random.choice(['foo', 'bar', 'ham']) for _ in range(n)],
        # a large column that would be expensive to read and won't compress well
        'c': [choices(string.ascii_letters, k=1000) for _ in range(n)]
    }
)

print(df.shape)

table = df.to_arrow()


for filesystem in ('', 'gs://'):
    uri = f'{filesystem}my-bucket/{uuid4()}'
    dt = deltalake.DeltaTable.create(uri, table.schema, partition_by=['a'], mode='overwrite')
    deltalake.write_deltalake(dt, table, mode='append')

    start = time()
    print(len(dt.files()))
    end = time()
    print(f'Time taken on {filesystem}: {(end - start) * 1000:.2f} ms')

    start = time()
    dt.to_pyarrow_dataset()
    end = time()
    print(f'Time taken on {filesystem}: {(end - start) * 1000:.2f} ms')
```

```
(1000, 2)
3
Time taken on : 0.05 ms
Time taken on : 0.81 ms
3
Time taken on gs://: 0.09 ms
Time taken on gs://: 287.64 ms
```

The fact that it takes _considerably_ longer to do on GCS indicates to me it is doing IO. And from what I understand of the description above it shouldn't be doing IO (all of that data is already loaded).

Further, if I run with `RUST_LOG=DEBUG` I see this coming from `to_pyarrow_dataset()`:

```
[2024-05-07T00:14:19Z DEBUG deltalake_core::table::builder] creating table builder with gs://my-bucket/9a861b4c-7244-43bb-a6e7-a9424864f385
[2024-05-07T00:14:19Z DEBUG deltalake_core::table::builder] build_storage() with gs://my-bucket/9a861b4c-7244-43bb-a6e7-a9424864f385
[2024-05-07T00:14:19Z DEBUG deltalake_core::table::builder] Loading a logstore based off the location: Url { scheme: "gs", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("my-bucket")), port: None, path: "/9a861b4c-7244-43bb-a6e7-a9424864f385", query: None, fragment: None }
[2024-05-07T00:14:19Z DEBUG deltalake_core::logstore] Found a storage provider for gs:// (gs://my-bucket/9a861b4c-7244-43bb-a6e7-a9424864f385)
[2024-05-07T00:14:20Z DEBUG deltalake_core::logstore] Found a logstore provider for gs://
```

So it sounds like it's creating a new DeltaTable or something?

**Reply by wjones127** [reply]
Yeah I'm not sure what these lines do these days:

https://github.com/delta-io/delta-rs/blob/e370d3457143cb77c3dfda212c677022bf630701/python/src/lib.rs#L870-L874

But maybe that could involve some IO right now. So perhaps it should be able to cache it? It might not right now for the sake of giving you the absolute latest version.

**Reply by ion-elgreco** [reply]
@wjones127 as far as I can see thats creating an iteration of the cached data in the state

**Reply by ion-elgreco** [reply]
@adriangb @wjones127 it's here: 

https://github.com/delta-io/delta-rs/blob/main/python%2Fdeltalake%2Ftable.py#L1097-L1099

There we build a new table instead of reusing the state. I'll see if I can make a fix

**Reply by ion-elgreco** [reply]
@adriangb can you try this one out: https://github.com/delta-io/delta-rs/pull/2485

**Reply by adriangb** [reply]
```
(1000, 2)
3
Time taken on : 0.07 ms
Time taken on : 2.83 ms
3
Time taken on gs://: 0.08 ms
Time taken on gs://: 0.99 ms
```

Great stuff!

**Reply by adriangb** [reply]
Would love a release with that fix 🙏🏻
